video/out/gpu/context.h

```diff
@@ -78,6 +78,11 @@ struct ra_swapchain_fns {
     // Gets the current framebuffer depth in bits (0 if unknown). Optional.
     int (*color_depth)(struct ra_swapchain *sw);
 
+    // Certain platforms may want to setup/check some things before actually
+    // drawing the next frame. If this exists and returns false, then subsequent
+    // start frame call will be skipped and the next frame will not be drawn.
+    bool (*wait_for_frame)(struct ra_swapchain *sw);
```

Comment:
Why expose this on the `ra_swapchain` level? Wouldn't it make more sense to just keep this in `ra_gl_ctx_params` (and friends) and have `ra_gl_ctx_start_frame` etc. call them directly?

Reply:
Also, calling it `wait_for_frame` in this context seems very wrong since it doesn't actually block. Maybe `check_visible` matches the actual implementation intent better?

In any case, I don't think this needs to be pulled up to the top-level API at all. You're writing a lot of redundant code and wrapping layers only to handle the logic at a higher level of abstraction than necessary.